### PR TITLE
[ADD] connector_dns

### PR DESCRIPTION
**Pull request by noahzaozao**

### CODE REVIEW

**Comment by noahzaozao** [comment]
@elicoidal I've moved connector_dns here,  please reveiw this PR.

File: connector_dns/README.rst
Comment (elicoidal) [inline connector_dns/README.rst]: I will create some content here

File: connector_dns/README.rst
Comment (elicoidal) [inline connector_dns/README.rst:60]: adapt

File: connector_dns/README.rst
Comment (elicoidal) [inline connector_dns/README.rst]: remove this comments

File: connector_dns/README.rst
Comment (elicoidal) [inline connector_dns/README.rst]: adapt

File: connector_dns/README.rst
Comment (elicoidal) [inline connector_dns/README.rst]: remove

File: connector_dns/dns.py
Comment (elicoidal) [inline connector_dns/dns.py]: add space after comma

File: connector_dns/dns.py
Comment (elicoidal) [inline connector_dns/dns.py]: idem

File: connector_dns/dns.py
Comment (elicoidal) [inline connector_dns/dns.py]: already be confirmed

File: connector_dns/dns.py
Comment (elicoidal) [inline connector_dns/dns.py]: Range: 1-20

File: connector_dns/dns.py
Comment (elicoidal) [inline connector_dns/dns.py]: Range: 1-604800

File: connector_dns/unit/backend_adapter.py
Comment (elicoidal) [inline connector_dns/unit/backend_adapter.py]: s/recorder/Record

File: connector_dns/README.rst
Comment (noahzaozao) [reply connector_dns/README.rst:60]: @elicoidal How to change this line ?

File: connector_dns/README.rst
Comment (noahzaozao) [reply connector_dns/README.rst]: @elicoidal remove the whole line ? 
- Odoo Community Association: `Icon <https://github.com/OCA/maintainer-tools/blob/master/template/module/static/description/icon.svg>`_.

File: connector_dns/__openerp__.py
Comment (lasley) [inline connector_dns/__openerp__.py]: Remove empty keys

File: connector_dns/connector.py
Comment (lasley) [inline connector_dns/connector.py]: Please isolate this from other backend types `dns_backend_id`

File: connector_dns/connector.py
Comment (lasley) [inline connector_dns/connector.py]: `models.AbstractModel`

File: connector_dns/connector.py
Comment (lasley) [inline connector_dns/connector.py]: I feel like this is going to clash with a lot of models. Is this necessary? Don't we have implicit state/exception monitoring with the job queue & whether we have a `dns_id`?

File: connector_dns/dns.py
Comment (lasley) [inline connector_dns/dns.py]: Please put this file in a `models` folder

File: connector_dns/dns.py
Comment (lasley) [inline connector_dns/dns.py]: We could probably go ahead and define the base DNS record type in the core; they're not really going to change between systems, so this will save a lot of repeated code downstream IMO

File: connector_dns/dns.py
Comment (lasley) [inline connector_dns/dns.py]: Sorry I am unclear on what a DNS Line is? Can we maybe get better naming, a help attribute, or default data for the list in order to help identify what this is for?

File: connector_dns/dns.py
Comment (lasley) [inline connector_dns/dns.py]: IMO might be good to index this col

File: connector_dns/tests/test_backend.py
Comment (lasley) [inline connector_dns/tests/test_backend.py]: You can remove this method

File: connector_dns/tests/test_backend.py
Comment (lasley) [inline connector_dns/tests/test_backend.py]: This seems named wrong?

File: connector_dns/tests/test_backend.py
Comment (lasley) [inline connector_dns/tests/test_backend.py]: This is also named wrong?

File: connector_dns/unit/backend_adapter.py
Comment (lasley) [inline connector_dns/unit/backend_adapter.py]: Remove blank lines

File: connector_dns/unit/binder.py
Comment (lasley) [inline connector_dns/unit/binder.py]: Is this parent class necessary? I haven't seen this in the other connectors.

Also, naming looks off.

File: connector_dns/unit/binder.py
Comment (lasley) [inline connector_dns/unit/binder.py]: Naming looks off, DNSPod

File: connector_dns/unit/binder.py
Comment (lasley) [inline connector_dns/unit/binder.py]: We should use `to_odoo`, particularly for new connectors

File: connector_dns/unit/export_synchronizer.py
Comment (lasley) [inline connector_dns/unit/export_synchronizer.py]: `_get_odoo_data`

**Comment by lasley** [comment]
Thanks for the submission @noahzaozao - comments inline. Also, can you please add test coverage? IMO this is going to be used as the base for a lot of things, so anything less than 80% is a 👎  for me - and honestly that's still pretty low.

FYI you may be able to copy a lot of testing logic from a [connector I am working on](https://github.com/laslabs/odoo-connector-carepoint/tree/release/9.0/connector_carepoint/tests). It has heavy coverage in areas where the other connectors are missing, like all of this core stuff. The [Magento connector](https://github.com/OCA/connector-magento/tree/9.0/magentoerpconnect/tests) has some good tests too. Hopefully those help 😄

File: connector_dns/dns.py
Comment (lasley) [inline connector_dns/dns.py]: Why redeclare? This is already in the abstract model

File: connector_dns/dns.py
Comment (lasley) [inline connector_dns/dns.py]: Typically the bind records are `_inherits` from regular Odoo records. I get that the regular record doesn't exist, but I think this is also breaking some core logic & not following the [interface spec](https://github.com/OCA/connector/blob/7dfc607ae31fbb2a0a4da5c3a4979401e75d185c/connector/backend_model.py#L69).

Note it says `openerp_id` in the spec, but I have been using `odoo_id` for new connectors. That's a design choice, but IMO it's best to deprecate anything named `openerp` for `odoo`

File: connector_dns/unit/binder.py
Comment (lasley) [inline connector_dns/unit/binder.py]: Here's some pesky [core logic break I was referencing](https://github.com/OCA/infrastructure-dns/pull/4/files#r76639007). Again, recommend `odoo_id` instead of `openerp_id` but we do need something.

**Comment by elicoidal** [comment]
Thanks @lasley for the valuable review!
We will review all the details and improve test coverage.

File: connector_dns/README.rst
Comment (elicoidal) [reply connector_dns/README.rst:60]: Don't worry.
I will create PR change the README once it is stable enough

File: connector_dns/README.rst
Comment (elicoidal) [reply connector_dns/README.rst]: Yes: we do not use the standard OCA logo so no need to give credit.
You can try to find a logo for the module but in this case you should give credit here...

**Comment by elicoidal** [comment]
Strange: we have added tests but the coverage remains the same...
@noahzaozao check the coveralls.io status per module

**Comment by lasley** [comment]
@elicoidal & @noahzaozao - I run into this a lot with mocks; it is likely one of two issues:
1. Not properly using context managers when mocking resources. There is a lot of unpatching involved with mocks as well. If these patches aren't removed properly, the object is orphaned from the original but still operates the same
2. When I try to get lazy and mock an object's superclass. Turns out this also orphans the object & messes with test results. I use a strategy [like this](https://github.com/laslabs/odoo-connector-carepoint/blob/release/9.0/connector_carepoint/tests/test_carepoint_exporter.py#L19) for supers.

The above said are purely hypothetical though. I'll take a look later today & after some coffee to see if anything points out.

**Comment by noahzaozao** [comment]
@lasley I'm concerned about change the change of 'to_openerp' to 'to_odoo' , current we're in odoo v8 , using the same rule of the OCA/connector module , e.g , in class Binder(ConnectorUnit) , not sure we'll have trouble with the change.

**Comment by elicoidal** [comment]
Odoo v8 is already officially "Odoo" branded.

we might have internally to perform a migration (but I dont think it 
makes sense to create a specific script in this module for that).

**Comment by lasley** [comment]
IMO it is best to make this conversion now vs later. As @elicoidal mentioned, v8 is already Odoo branded - just kind of halfway. This leaves the design choice up to you as a developer.

The reason I recommend making the change now is due to the fact that this module will be the basis of all DNS modules to come. The change has to be made at some point, but making it later also forces a change in all reliant modules. 

You are right though that the upstream module uses `to_openerp` still, even [in v9](https://github.com/OCA/connector/blob/9.0/connector/connector.py#L357). I just aliased it, calling the `super.to_openerp` in the `to_odoo` instead of `super.to_odoo`. I can guarantee you if I have a hand in the v10 upgrade of connector, this change will be made in core. All `openerp` branding is deprecated and should be removed.

Note that if you alias like recommended, there is one change in the center of your app that needs to be made. If you don't, we're talking a diff that spans more than half the files.

**Comment by elicoidal** [comment]
@lasley
Can we move forward and merge this PR without full coverage?
UT here are very complex and would require quite some time, which could be achieve in the future.
Any help on improving the coverage would be welcome but I think it can be done gradually after the first merge

**Comment by lasley** [comment]
The base of the connector module sets up all the common test framework for all the other modules. If we merge this as is, we are guaranteeing that we cannot test any of the other modules either. 

60% of this is tested, which means 40% is a ticking time bomb IMO. We plan on building more apps on top of this timebomb with the intent of managing essentially the most important part of the web stack - DNS. I definitely can't take the risk using this in my infrastructure, so I can't provide a thumb to recommend anyone else use it on their's.

That said, I'll see what I can do about raising the coverage on my end and submitting a PR to your branch. I'll keep you posted 😄

**Comment by elicoidal** [comment]
@lasley @pedrobaeza can we move forward here?

**Comment by gurneyalex** [comment]
:+1: (no test)

**Comment by elicoidal** [comment]
@lasley I understand and agree with the overall objective: if you can help that is welcome because we lack experience in testing connector.

**Comment by lasley** [comment]
I'm going to help sorry I've just been backed up. On my list of things to do

File: connector_dns/models/dns.py
Comment (lasley) [inline connector_dns/models/dns.py]: What is this attribute used for?

**Comment by lasley** [comment]
@noahzaozao @elicoidal - FYI I started [my branch](https://github.com/laslabs/infrastructure-dns/tree/release/8.0/connector-dns) and will have a PR soon. I am restructuring and adding in a lot of missing pieces, so it may take a bit.

Can you please squash your existing commits so we have a cleaner branch for me to rebase onto?

File: connector_dns/models/dns.py
Comment (lasley) [inline connector_dns/models/dns.py]: IMO it would be better to use the same bind strategy as all of the other connector modules, using `_inherits` and a specific bind model e.g `aws.dns.domain`. 

Otherwise, these domains/records cannot be synced to multiple systems at the same time. This also conflicts with a lot of the documentation, even within this module.

I will be making this proposal/change as part of my PR, I just wanted to bring it up here in case there was a specific reasoning behind this.

Comment (elicoidal) [reply connector_dns/models/dns.py]: @lasley Nothing specific. The base was started 2 years ago while we didnot manage so well the techno and we did not rebase everything lately (too busy ...).
Your help is welcome on the matter and we can participate on the new PR

Comment (lasley) [reply connector_dns/models/dns.py]: Excellent I was hoping that was the case, thanks @elicoidal - I'm going over everything with a fine tooth comb as I'm adding these tests. Making sure we've got what we need for scalability and all that jazz 😄

Comment (elicoidal) [reply connector_dns/models/dns.py]: @lasley Hope you are fine: did you make any progress here?

Comment (lasley) [reply connector_dns/models/dns.py]: Yup- I'll push something today. I was waiting for a response back on what the `line` column in `dns.record` was for as well if you could help out with that? It seems redundant with other cols, but also maybe looks like it could have been the beginning of a record versioning feature?

File: connector_dns/models/dns.py
Comment (lasley) [inline connector_dns/models/dns.py]: Was there a reason for leaving out the DNS record types? These are pretty nailed in stone for the most part - A record, etc. Am I misunderstanding this col/computation?

File: connector_dns/models/dns.py
Comment (lasley) [inline connector_dns/models/dns.py]: Will you mind if I change this to DnsZone? Domain is ambiguous with the actual domains that are being hosted.

File: connector_dns/models/dns.py
Comment (elicoidal) [reply connector_dns/models/dns.py]: We have it implemented in the dnspod connector: https://github.com/OCA/l10n-china/pull/22/files#diff-f3401e9b4eb95d57f90bce3b630d928cR141

It is not only the typical DNS type (A CNAME etc) but also the ones actually accepted by the provider. I have to reckon that the current treatment by dnspod is not exactly the best option...

File: connector_dns/models/dns.py
Comment (elicoidal) [reply connector_dns/models/dns.py]: in dnspod (https://github.com/OCA/l10n-china/pull/22/files#diff-f3401e9b4eb95d57f90bce3b630d928cR148): represent the carrier (China Telecom, Unicom, etc.).
I think it could be dropped from the main model.

File: connector_dns/models/dns.py
Comment (elicoidal) [reply connector_dns/models/dns.py]: Not sure here (just don't know what is acceptable terminology). We would have to migrate our records but that is no big deal if there is strong argument.

File: connector_dns/models/dns.py
Comment (lasley) [reply connector_dns/models/dns.py]: Awesome thanks, I was so confused! Dropped.

File: connector_dns/models/dns.py
Comment (lasley) [reply connector_dns/models/dns.py]: Alright sweet - I added all the default record types then, and we can still leave the method for custom stuff just in case.

I'm wondering if we should instead make a model for `dns.record.type` and use that instead of a select. In this way, we can define what custom connector the type is for. That would also allow us to add help attributes, etc. Thoughts?

File: connector_dns/models/dns.py
Comment (lasley) [reply connector_dns/models/dns.py]: Hmmmm if you're using this in production maybe we leave a change like that out in order to avoid migrations. I changed the `openerp_id` to `odoo_id` too - with the specific intent to ease future migrations, but I didn't realize you guys are live already. I can reverse the breaking fixes.

Comment (lasley) [reply connector_dns/models/dns.py]: Most DNS providers and systems do use the terminology zone though:
- https://en.wikipedia.org/wiki/DNS_zone
- https://technet.microsoft.com/en-us/library/cc753643(v=ws.11).aspx
- http://docs.aws.amazon.com/Route53/latest/DeveloperGuide/CreatingHostedZone.html
- http://searchsoa.techtarget.com/definition/zone-file

File: connector_dns/models/dns.py
Comment (lasley) [inline connector_dns/models/dns.py]: Is this being used? I can't seem to find any sort of corresponding logic here. Did this instead get implemented as `state` in the abstract binding?

File: connector_dns/connector.py
Comment (lasley) [inline connector_dns/connector.py]: How important is this column to your internal business logic? Reason I ask is that I am finding that it is somewhat ambiguous in the following examples, where using the `sync_date` as the status would eliminate the ambiguity:
- Record is synced to remote system. It is then changed on the remote system. Record is out of sync w/ connector, while being indicated that sync is done
- Record is synced to remote system. It is then changed in Odoo, resynced, and failed. There is now no indication that the record originally passed sync

I do think I like the idea of knowing about failures though. What if we added:
- `fail_date` - this column is the inverse of `sync_date`, providing the last sync fail time

File: connector_dns/connector.py
Comment (elicoidal) [inline connector_dns/connector.py]: modify the comment

File: connector_dns/connector.py
Comment (elicoidal) [reply connector_dns/connector.py]: Yes this is about the domain synchronization failure. OK for fail date (nice to have as if sync_date is blank then this s good enough IMO)

File: connector_dns/models/dns.py
Comment (elicoidal) [reply connector_dns/models/dns.py]: If I am not mistaken it is used to validate the connection to the DNS provider (check everything is alright: ID/password)

File: connector_dns/models/dns.py
Comment (elicoidal) [reply connector_dns/models/dns.py]: OK with the change. We will migrate anyway when going to v9 ;) so better to have a strong framework than keeping legacy bad constructions

File: connector_dns/models/dns.py
Comment (elicoidal) [reply connector_dns/models/dns.py]: Is it possible that we have to sync the record  type with the provider? 
I see the benefit of a model (more flexible for example to adapt with DNSpod specific values)

File: connector_dns/models/dns.py
Comment (elicoidal) [reply connector_dns/models/dns.py]: @lasley Thanks for your help! Once the model is stabilized, we will adapt the DNSPod module

**Comment by lasley** [comment]
@elicoidal @noahzaozao - I'm done refactoring & in the process of adding the tests now. Here's the diff if you want to take a look https://github.com/noahzaozao/infrastructure-dns/compare/8.0-add-connector-dns...laslabs:release/8.0/connector-dns

When done am I submitting a PR to your branch, or to OCA?

Here's an accounting of my changes:

```
* Separate models by file
* Add `import_synchronizer`
* Add `export_synchronizer`
* Add `delete_synchronizer`
* Add `mapper`
* Add `fail_date` to abstract bind, remove state
* Add external date field handling to binder
* Rename `dns.domain` to `dns.zone`
* Remove `line` field from `dns.record`
* Add security
* Code cleanup
* Add test coverage
* Don't cover test assets
* Change bind field to `odoo_id` to ease future migration
```

File: connector_dns/models/dns.py
Comment (lasley) [reply connector_dns/models/dns.py]: Yup I don't see why we can't bind on types as well. Will make that change

**Comment by elicoidal** [comment]
Better to @noahzaozao branch I think as this was the initiator.

**Comment by lasley** [comment]
Sounds good @elicoidal 

@noahzaozao can you squash the commits in your branch please?

**Comment by noahzaozao** [comment]
@lasley I've squashed the commits in my branch , but seems now I've conflict your PR.

**Comment by elicoidal** [comment]
@lasley what is the situation on this PR?

**Comment by lasley** [comment]
@elicoidal - I'm 👍  on it, but it's basically all my code at this point so IMO my thumb doesn't really count for much. Looks like there is a Runbot error - I haven't looked into resolving it, but can if you would like. It's not erring on mine though, so I would need to submit a new PR in order to properly troubleshoot.

**Comment by elicoidal** [comment]
I have reviewed the original PR and :+1: but this is highly technical to me and it would require some functional tests (which is tricky as this is only the technical layer).
@guewen @gurneyalex @hbrunn if you have the opportunity to review this work?

**Comment by elicoidal** [comment]
BTW, something I dont understand: runbot appears red but there is no log and the db is accessible.
Something I am missing?

**Comment by elicoidal** [comment]
Functional test
- Click on Zone menu/create and get the following error: Error: Unknown field state in domain [["state","not in",["done"]]]
- Click on Backend menu/create and get the following error: Error: Unknown field state in domain [["state","not in",["draft","exception"]]]
- It is strange not to have the Zone mandatory in the DNS record
- Title of the Backend menu could be improved (Currently "DNS Backend Action"). Same for Records.
- Settings/Connector menu is empty

**Comment by hbrunn** [comment]
runbot is red because of this warning:  WARNING server  openerp.models:4119 create ir.actions.act_window.create() with unknown fields: type_id

**Comment by noahzaozao** [comment]
@elicoidal Not sure how to fix the runbot warning

**Comment by elicoidal** [comment]
@lasley do you have any idea here how to get this rolling?

**Comment by lasley** [comment]
Flying to Belgium today, was planning on knocking this & prelim Docker connector out on the plane.

**Comment by lasley** [comment]
All points attended to except for `Settings/Connector menu is empty`. Forgive my ignorance on this one, but what exactly should go in that menu? 

noahzaozao/infrastructure-dns#4

**Comment by hbrunn** [comment]
@lasley usually, a wizard derived from `ir.config.settings` that allows users to configure the module. There might (I'm not very active with the connector framework) even be some base module where you can inject a checkbox to install your module from the configuration wizard

**Comment by lasley** [comment]
Exactly what I needed, thanks @hbrunn 

Will submit a new PR adding the config settings sometime this week

**Comment by lasley** [comment]
Added blank config view in noahzaozao/infrastructure-dns#5 - there was no need for the model at the moment because no extra modules to enable.

**Comment by noahzaozao** [comment]
@elicoidal Once this PR has been merged , I can start re-factory the dependency modules such as connector_dnspod

**Comment by noahzaozao** [comment]
@elicoidal Shall we move forward with this PR ?

**Comment by elicoidal** [comment]
@lasley Are we good to move forward?

**Comment by lasley** [comment]
Yeah it would be nice to see this merged

**Comment by elicoidal** [comment]
@hbrunn @gurneyalex any feedback before I merge?

**Comment by hbrunn** [comment]
I didn't do any review, so I don't have anything to say about this PR (and it's too big to slip in a review right now). Better rely on people who actually looked at the code for a judgement, formally nobody approved it up to now...

**Comment by elicoidal** [comment]
@hbrunn Thanks for the quick feedback. I agree with you on both remarks :)

On my side:  :+1: 
@lasley ?

**Review by lasley — APPROVED** [review]
👍  from me on the stuff I didn't write. Obviously a 👍  on the stuff I did 😉

**Comment by flotho** [comment]
any chance to see this PR merge ?

**Comment by elicoidal** [comment]
I ll merge it!

**Comment by elicoidal** [comment]
@noahzaozao @seb-elico

**Pull request merged**
